compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp

```diff
@@ -0,0 +1,227 @@
+      sharedMemAllocs.push_back(alloc);
+    }
+  }
+  if (sharedMemAllocs.size() < 2) {
+    return failure();
+  }
+
+  for (auto alloc : sharedMemAllocs) {
+    LivenessRange livenessRange =
+        getLivenessRange(alloc, dominanceInfo, rootOp->getRegion(0));
+    livenessMap.insert(std::make_pair(alloc, livenessRange));
+  }
+  return success();
+}
+
+/// Helper to determine if 2 liveness ranges overlap. The ranges overlap if
+/// the `begin` of one of the ranges falls on or between the `begin` and `end`
+/// of the other range.
+static bool livenessRangesOverlap(LivenessRange range, LivenessRange other,
+                                  DominanceInfo &dominanceInfo,
+                                  Block *entryBlock) {
+  return (dominanceInfo.dominates(range.first, other.first) &&
+          dominanceInfo.dominates(other.first, range.second)) ||
+         (dominanceInfo.dominates(other.first, range.first) &&
+          dominanceInfo.dominates(range.first, other.second));
+}
+
+/// Returns the sets of allocations that have overlapping liveness ranges.
+/// Each set will not have any overlapping liveness ranges with any other set.
+static SmallVector<AliasGroup>
+computeAliasGroups(DenseMap<Operation *, LivenessRange> livenessMap,
+                   DominanceInfo &dominanceInfo, Block *entryBlock) {
+  SmallVector<AliasGroup> aliasGroups;
+  for (auto val : livenessMap) {
+    Operation *alloc = val.getFirst();
+    LivenessRange livenessRange = val.getSecond();
+    // Check for overlapping liveness with other alias groups. Combine all
+    // groups that overlap with the `alloc`.
+    SmallVector<AliasGroup> overlappingGroups, newAliasGroups;
+    for (auto group : aliasGroups) {
+      if (llvm::any_of(group, [&](Operation *alias) {
+            return livenessRangesOverlap(livenessRange, livenessMap[alias],
+                                         dominanceInfo, entryBlock);
+          })) {
+        overlappingGroups.push_back(group);
+      } else {
+        newAliasGroups.push_back(group);
+      }
+    }
+    AliasGroup combinedGroup;
+    for (auto group : overlappingGroups) {
+      combinedGroup.append(group);
+    }
+    combinedGroup.push_back(alloc);
+    newAliasGroups.push_back(combinedGroup);
+    aliasGroups = newAliasGroups;
+  }
+  return aliasGroups;
+}
+
+/// Get all subviews of the `op` recursively. Child subviews come before their
+/// parent views in the set of `views`.
+static void getAllSubviews(Operation *op, SetVector<Operation *> &views) {
+  for (auto user : op->getUsers()) {
+    if (isa<memref::SubViewOp>(user)) {
+      if (user->getUsers().empty()) {
+        return;
+      }
+      getAllSubviews(user, views);
+      views.insert(user);
+    }
+  }
+}
+
+/// Sink all share memory allocations and their subviews within the CFG, so
+/// they are as close as possible to their first user.
+static void sinkSharedMemoryAllocsInCFG(FunctionOpInterface funcOp) {
```

Comment:
Why are you doing this movement? Shared memory is a property of the kernel, so the location of the `memref.alloc` does not matter at all later on.

Reply:
This was because of the util I was using to place barriers. I rewrote it without the util so I don't need this anymore. It also removes the LLVMGPU dependency.